File: security/src/main/java/ru/virgil/security/SecurityConfig.java
package ru.virgil.security;

import lombok.RequiredArgsConstructor;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpMethod;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;
import org.springframework.web.cors.UrlBasedCorsConfigurationSource;
import ru.virgil.security.entity.AuthMethods;
import ru.virgil.security.firebase.FirebaseAuthenticationFilter;
import ru.virgil.security.firebase.FirebaseAuthorizationProvider;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.Optional;


@RequiredArgsConstructor
@Configuration
public class SecurityConfig extends WebSecurityConfigurerAdapter {

    public static final Map<AuthMethods, String> AUTH_API_PATHS = Map.of(
            AuthMethods.FIREBASE, "/auth/firebase"
    );
    public static final Map<AuthMethods, String> AUTH_PAGE_PATHS = Map.of(
            AuthMethods.FIREBASE, "/auth/firebase/page/**"
    );
    private final FirebaseAuthorizationProvider firebaseAuthorizationProvider;
    private final AuthenticationEntryPoint authenticationEntryPoint;
    private final SecurityProperties securityProperties;

    @Override
    protected void configure(HttpSecurity httpSecurity) throws Exception {
        String[] propertyIgnoredPaths = Optional.ofNullable(securityProperties.anonymousPaths()).orElse(new String[0]);
        httpSecurity
                // todo: стандартный редирект на страницу успешной безопасности
                // todo: разобраться, как включить
                .cors()
                .and()
                .csrf().disable()
                .authorizeRequests()
                .mvcMatchers(AUTH_PAGE_PATHS.values().toArray(String[]::new)).permitAll()
                .mvcMatchers("/", "/favicon.ico", "/error").permitAll()
                .mvcMatchers(propertyIgnoredPaths).permitAll()
                .mvcMatchers("/**").authenticated()
                .and()
                .addFilterBefore(new FirebaseAuthenticationFilter(authenticationManager()),
                        UsernamePasswordAuthenticationFilter.class)
                .authenticationProvider(firebaseAuthorizationProvider)
        // todo: разобраться, как лучше реагировать на ошибки и не подставлять безопасность
        // .exceptionHandling()
        // .authenticationEntryPoint(authenticationEntryPoint)
        ;
    }

    @Bean
    protected CorsConfigurationSource corsConfigurationSource() {
        CorsConfiguration authCorsConfiguration = buildAuthCorsConfiguration();
        CorsConfiguration resultCorsConfiguration = buildResultCorsConfiguration();
        UrlBasedCorsConfigurationSource configurationSource = new UrlBasedCorsConfigurationSource();
        configurationSource.registerCorsConfiguration(AUTH_API_PATHS.get(AuthMethods.FIREBASE), authCorsConfiguration);
        configurationSource.registerCorsConfiguration("/**", resultCorsConfiguration);
        return configurationSource;
    }

    private CorsConfiguration buildResultCorsConfiguration() {
        CorsConfiguration resultCorsConfiguration = new CorsConfiguration();
        resultCorsConfiguration.setAllowedOrigins(List.of("http://localhost:4200"));
        resultCorsConfiguration.setAllowedHeaders(List.of(HttpHeaders.AUTHORIZATION, HttpHeaders.COOKIE));
        resultCorsConfiguration.setAllowedMethods(Arrays.stream(HttpMethod.values()).map(Enum::name).toList());
        resultCorsConfiguration.setAllowCredentials(true);
        return resultCorsConfiguration;
    }

    private CorsConfiguration buildAuthCorsConfiguration() {
        CorsConfiguration authCorsConfiguration = new CorsConfiguration();
        authCorsConfiguration.setAllowedOrigins(List.of("http://localhost:4200"));
        authCorsConfiguration.setAllowedMethods(List.of(HttpMethod.POST.name()));
        authCorsConfiguration.setAllowedHeaders(List.of(HttpHeaders.AUTHORIZATION));
        authCorsConfiguration.setExposedHeaders(List.of(HttpHeaders.AUTHORIZATION));
        authCorsConfiguration.setAllowCredentials(true);
        return authCorsConfiguration;
    }
}

File: security/src/main/java/ru/virgil/security/SecurityConfig.kt
package ru.virgil.security

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpMethod
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource
import ru.virgil.security.entity.AuthMethods
import ru.virgil.security.firebase.FirebaseAuthenticationFilter
import ru.virgil.security.firebase.FirebaseAuthorizationProvider
import java.util.*
import java.util.Map
import kotlin.collections.List
import kotlin.collections.listOf
import kotlin.collections.toTypedArray


@Configuration
class SecurityConfig(
    private val firebaseAuthorizationProvider: FirebaseAuthorizationProvider,
    private val authenticationEntryPoint: AuthenticationEntryPoint,
    private val securityProperties: SecurityProperties,
    private val objectMapper: ObjectMapper,
) : WebSecurityConfigurerAdapter() {

    @Throws(Exception::class)
    override fun configure(httpSecurity: HttpSecurity) {
        val propertyIgnoredPaths: List<String> = securityProperties.anonymousPaths
        httpSecurity // todo: стандартный редирект на страницу успешной безопасности
            // todo: разобраться, как включить
            .cors()
            .and()
            .csrf().disable()
            .authorizeRequests()
            .mvcMatchers(*AUTH_PAGE_PATHS.values.toTypedArray()).permitAll()
            .mvcMatchers("/", "/favicon.ico", "/error").permitAll()
            .mvcMatchers(*propertyIgnoredPaths.toTypedArray()).permitAll()
            .mvcMatchers("/**").authenticated()
            .and()
            .addFilterBefore(
                FirebaseAuthenticationFilter(authenticationManager(), securityProperties, objectMapper),
                UsernamePasswordAuthenticationFilter::class.java
            )
            .authenticationProvider(firebaseAuthorizationProvider) // todo: разобраться, как лучше реагировать на ошибки и не подставлять безопасность
        // .exceptionHandling()
        // .authenticationEntryPoint(authenticationEntryPoint)
    }

    @Bean
    protected fun corsConfigurationSource(): CorsConfigurationSource? {
        val authCorsConfiguration = buildAuthCorsConfiguration()
        val resultCorsConfiguration = buildResultCorsConfiguration()
        val configurationSource = UrlBasedCorsConfigurationSource()
        configurationSource.registerCorsConfiguration(AUTH_API_PATHS[AuthMethods.FIREBASE]!!, authCorsConfiguration)
        configurationSource.registerCorsConfiguration("/**", resultCorsConfiguration)
        return configurationSource
    }

    private fun buildResultCorsConfiguration(): CorsConfiguration {
        val resultCorsConfiguration = CorsConfiguration()
        resultCorsConfiguration.allowedOrigins = listOf(ORIGIN)
        resultCorsConfiguration.allowedHeaders = listOf(X_AUTH_TOKEN_HEADER, HttpHeaders.CONTENT_TYPE)
        resultCorsConfiguration.allowedMethods = HttpMethod.values().map { obj: HttpMethod -> obj.name }
        return resultCorsConfiguration
    }

    private fun buildAuthCorsConfiguration(): CorsConfiguration {
        val authCorsConfiguration = CorsConfiguration()
        authCorsConfiguration.allowedOrigins = listOf(ORIGIN)
        authCorsConfiguration.allowedMethods = listOf(HttpMethod.POST.name)
        authCorsConfiguration.allowedHeaders = listOf(HttpHeaders.AUTHORIZATION)
        authCorsConfiguration.exposedHeaders = listOf(X_AUTH_TOKEN_HEADER)
        return authCorsConfiguration
    }

    companion object {

        const val ORIGIN = "http://localhost:4200"
        const val X_AUTH_TOKEN_HEADER = "X-Auth-Token"

        @JvmField
        val AUTH_API_PATHS = Map.of(
            AuthMethods.FIREBASE, "/auth/firebase"
        )

        @JvmField
        val AUTH_PAGE_PATHS = Map.of(
            AuthMethods.FIREBASE, "/auth/firebase/page/**"
        )
    }
}
